### PR TITLE
[Order] Add trailing stop implementation

### PR DESCRIPTION
**Pull request by Herklos**

### CODE REVIEW

File: octobot_trading/__init__.py [hunk 1]
```diff
@@ -15,4 +15,4 @@
 #  License along with this library.
 
 PROJECT_NAME = "OctoBot-Trading"
-VERSION = "1.7.2"  # major.minor.revision
+VERSION = "1.8.0"  # major.minor.revision
```
Comment (GuillaumeDSM) [inline octobot_trading/__init__.py:18]: :tada:

File: octobot_trading/orders/types/trailing/trailing_stop_order.pxd [hunk 2]
```diff
@@ -20,4 +20,16 @@ from octobot_trading.data.order cimport Order
 
 cdef class TrailingStopOrder(Order):
     cdef object trailing_stop_price_hit_event # object is asyncio.Event
-    cdef object wait_for_hit_event_task # object is asyncio.Task
+    cdef object trailing_price_hit_event # object is asyncio.Event
+    cdef object wait_for_stop_price_hit_event_task # object is asyncio.Event
+    cdef object wait_for_price_hit_event_task # object is asyncio.Event
```
Comment (GuillaumeDSM) [inline octobot_trading/orders/types/trailing/trailing_stop_order.pxd:25]: excellent idea to add types in comments

File: octobot_trading/orders/types/trailing/trailing_stop_order.py [hunk 3]
```diff
@@ -16,40 +16,156 @@
...
-                add_event(self.origin_price, self.creation_time, self.side is TradeOrderSide.SELL)
+        await self._reset_events(self.origin_price, self.creation_time, force_reset=False)
+
+    async def set_trailing_percent(self, trailing_percent):
+        """
+        Set trailing percent and reset event and tasks
+        :param trailing_percent: the new trailing percent
+        """
+        self.trailing_percent = trailing_percent
+        await self._reset_events(self.origin_price, self.creation_time, force_reset=True)
```
Comment (GuillaumeDSM) [inline octobot_trading/orders/types/trailing/trailing_stop_order.py:49]: is it OK to call a cythonized positional argument as a named argument ? I remember we had issues about this

File: tests/orders/__init__.py [hunk 4]
```diff
@@ -77,7 +78,21 @@ def take_profit_buy_order(event_loop, simulated_trader):
     return TakeProfitOrder(trader_instance, side=TradeOrderSide.BUY)
 
 
-@pytest.fixture()
+@pytest.yield_fixture()
 def take_profit_limit_order(event_loop, simulated_trader):
+    _, exchange_manager, trader_instance = simulated_trader
+    yield TakeProfitLimitOrder(trader_instance)
+    exchange_manager.exchange_personal_data.orders_manager.clear()
```
Comment (GuillaumeDSM) [inline tests/orders/__init__.py:85]: why having to clear it before exchange stop ?

File: tests/orders/types/trailing/test_trailing_stop_order.py [hunk 5]
```diff
@@ -13,4 +13,118 @@
...
+    order.exchange_manager.is_backtesting = True  # force update_order_status
+    await order.initialize()
+    price_events_manager = order.exchange_manager.exchange_symbols_data.get_exchange_symbol_data(
+        DEFAULT_SYMBOL_ORDER).price_events_manager
+    order.exchange_manager.exchange_personal_data.orders_manager.upsert_order_instance(order)
+    return order, order_price, price_events_manager
+
+
+def get_price_percent(price, percent):
+    return price * (1 - percent / 100)
```
Comment (GuillaumeDSM) [inline tests/orders/types/trailing/test_trailing_stop_order.py:130]: 👍 
Should we also test other side's trailing stops ?

File: octobot_trading/orders/types/trailing/trailing_stop_order.py (hunk 3 above)
Comment (Herklos) [reply octobot_trading/orders/types/trailing/trailing_stop_order.py:49]: Hum... I don't remember, I'll have a look at this method in compiled version

File: tests/orders/types/trailing/test_trailing_stop_order.py (hunk 5 above)
Comment (Herklos) [reply tests/orders/types/trailing/test_trailing_stop_order.py:130]: Yes we should :+1:

File: tests/orders/__init__.py (hunk 4 above)
Comment (Herklos) [reply tests/orders/__init__.py:85]: To kill the order related price_event tasks to prevent CancelledErrors

Comment (GuillaumeDSM) [reply tests/orders/__init__.py:85]: I mean exchange_manager.clear() should already call this, it's not working ?

Comment (Herklos) [reply tests/orders/__init__.py:85]: The problem here is not the initial `TakeProfitLimitOrder` clear(). But it's artificial order when created that is added in `exchange_manager.exchange_personal_data.orders_manager `orders list.

File: tests/orders/types/trailing/test_trailing_stop_order.py (hunk 5 above)
Comment (Herklos) [reply tests/orders/types/trailing/test_trailing_stop_order.py]: Added one named `test_trailing_stop_with_new_price_inversed` :wink:

File: octobot_trading/orders/types/trailing/trailing_stop_order.py (hunk 3 above)
Comment (Herklos) [reply octobot_trading/orders/types/trailing/trailing_stop_order.py]: `force_reset` doesn't exist anymore :wink:

Comment (GuillaumeDSM) [reply octobot_trading/orders/types/trailing/trailing_stop_order.py]: perfect :)

**Pull request merged**